runners/spark/src/main/java/org/apache/beam/runners/spark/util/CachedSideInputReader.java

```diff
@@ -17,17 +17,31 @@
  */
 package org.apache.beam.runners.spark.util;
 
-import java.util.HashMap;
-import java.util.Map;
-import java.util.Objects;
+import com.google.common.cache.Cache;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
 import javax.annotation.Nullable;
 import org.apache.beam.runners.core.SideInputReader;
+import org.apache.beam.runners.spark.util.SideInputStorage.Key;
+import org.apache.beam.runners.spark.util.SideInputStorage.Value;
 import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
 import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.spark.util.SizeEstimator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /** {@link SideInputReader} that caches materialized views. */
 public class CachedSideInputReader implements SideInputReader {
 
+  private static final Logger LOG = LoggerFactory.getLogger(CachedSideInputReader.class);
+
+  /**
+   * Keep references for the whole lifecycle of CachedSideInputReader otherwise sideInput needs to
+   * be de-serialized again.
+   */
+  private Set<?> sideInputReferences = new HashSet<>();
```

Comment:
Aren't values ever removed from here? Or I am misreading this one, seems like it can overflow and even prevent `SideInputStorage` from being GCed.

Reply:
references removed because different solution was used. [more](https://github.com/apache/beam/pull/7091#issuecomment-451961840)